integration/yarn/src/test/java/alluxio/yarn/ClientTest.java

```diff
@@ -52,45 +52,111 @@ public void before() {
     PowerMockito.mockStatic(YarnClient.class);
     mYarnClient = (YarnClient) Mockito.mock(YarnClient.class);
     Mockito.when(YarnClient.createYarnClient()).thenReturn(mYarnClient);
+
+    Configuration.set(PropertyKey.INTEGRATION_MASTER_RESOURCE_MEM, "2048.00MB");
+    Configuration.set(PropertyKey.INTEGRATION_MASTER_RESOURCE_CPU, "4");
+    Configuration.set(PropertyKey.INTEGRATION_WORKER_RESOURCE_MEM, "2048.00MB");
+    Configuration.set(PropertyKey.WORKER_MEMORY_SIZE, "4096.00MB");
+    Configuration.set(PropertyKey.INTEGRATION_WORKER_RESOURCE_CPU, "8");
   }
 
-  /**
-   * Tests that the main method properly reads args.
-   */
-  @Test(timeout = 10000)
-  public void mainTest() throws Exception {
-    Client mockClient = spy(new Client());
-    Mockito.doNothing().when(mockClient).run();
-    PowerMockito.whenNew(Client.class)
-        .withNoArguments()
-        .thenReturn(mockClient);
-
-    Client.main(new String[] {
-        "-resource_path", "hdfs://test",
-        "-am_memory", "1024",
+  @Test(timeout = 1000)
+  public void notEnoughMemoryForApplicationMaster() throws Exception {
+    int appMasterMem = 1024;
+    Resource resource = Resource.newInstance(appMasterMem / 2, 4);
+    generateMaxAllocation(resource);
+    mThrown.expect(RuntimeException.class);
+    mThrown.expectMessage(ExceptionMessage.YARN_NOT_ENOUGH_RESOURCES.getMessage(
+        "ApplicationMaster", "memory", appMasterMem, resource.getMemory()));
+    Client client = new Client();
+    client.parseArgs(new String[] {
+        "-resource_path", "test",
+        "-am_memory", Integer.toString(appMasterMem),
         "-am_vcores", "2"});
-    PowerMockito.verifyNew(Client.class).withNoArguments();
+    client.run();
+  }
+
+  @Test(timeout = 1000)
+  public void notEnoughVCoreForApplicationMaster() throws Exception {
+    int appMasterMem = 1024;
+    int appMasterCore = 2;
+    Resource resource = Resource.newInstance(appMasterMem, appMasterCore / 2);
+    generateMaxAllocation(resource);
+    mThrown.expect(RuntimeException.class);
+    mThrown.expectMessage(ExceptionMessage.YARN_NOT_ENOUGH_RESOURCES.getMessage(
+        "ApplicationMaster", "virtual cores", appMasterCore, resource.getVirtualCores()));
+    Client client = new Client();
+    client.parseArgs(new String[] {
```

Comment:
I see what you mean. Could we instead make parseArgs public, and possibly rename it to init? It's basically acting as an extension of the constructor. Alternately, we could pass the `String[] args` to the `Client` constructor. Generally we try to avoid modifying visibility levels just so that test can call things.